docs/DifferentiableTypes.md

```diff
@@ -0,0 +1,289 @@
+
+<p align="center">
+  <img src="https://upload.wikimedia.org/wikipedia/commons/3/37/Pushforward.svg" align=center>
+  <br>
+  <sub>Image showing two differentiable manifolds: a sphere and a spheroid.</sub>
+  <br>
+  <sub>From https://en.wikipedia.org/wiki/Pushforward_(differential).</sub>
+</p>
+
+Here are the important points about the `Differentiable` protocol:
+* `associatedtype TangentVector`: represents the type of directional derivatives computed via forward-mode differentiation.
+* `associatedtype CotangentVector`: represents the type of gradient values computed via reverse-mode differentiation.
+* `CotangentVector` types are used and produced by differential operators like `gradient` and `pullback`.
+* `var allDifferentiableVariables: AllDifferentiableVariables`: represents all differentiable variables in an instance of the conforming type, where `AllDifferentiableVariables` is the type of all differentiable variables.
+  * The motivation/design behind this is related to key path iteration .
+* `TangentVector`, `CotangentVector`, and `AllDifferentiableVariables` are closely related.
+  * All three associated types must themselves conform to `Differentiable`.
+  * The `Differentiable` protocol associated types of the associated types themselves are defined to be mathematically correct.
+    * `Foo.TangentVector.TangentVector` is `Foo.TangentVector` itself.
+    * `Foo.CotangentVector.TangentVector` is `Foo.CotangentVector` itself.
+    * `Foo.TangentVector.CotangentVector` is `Foo.CotangentVector`.
+    * `Foo.CotangentVector.CotangentVector` is `Foo.TangentVector`.
+    * `Foo.AllDifferentiableVariables` has the same `TangentVector` and `CotangentVector` as `Foo`.
+  * Additionally, `TangentVector` and  `CotangentVector` must conform to `AdditiveArithmetic`, so that they can be zero-initialized and accumulated via addition. These are necessary to perform the chain rule of differentiation.
+* Manifold operations.
+  * These currently involve `func tangentVector(from cotangent: CotangentVector)` and `func moved(along cotangent: CotangentVector`). These operations can be useful for implementing manifold-related algorithms, like optimization on manifolds, but are not relevant for simple differentiation use cases.
+
+The standard library defines conformances to the `Differentiable` protocol for `Float` and `Double`, as well as a conditional conformance for `Tensor`:
+
+```swift
+extension Float: Differentiable {
+    public typealias TangentVector = Float
+    public typealias CotangentVector = Float
+    public typealias AllDifferentiableVariables = Float
+}
+// The conformance for `Double` is defined similarly.
+
+extension Tensor: Differentiable where Scalar : TensorFlowFloatingPoint {
```

Comment:
Not sure why we should mention `Tensor` here. Can we define a new `Vector` type to demonstrate this?

Reply:
I provided more context for `Tensor` (stated that it's from the TensorFlow library) and linked to API docs. I think this is sufficient.